File: Java/com/leetcode/validSudoku/ValidSudoku.java
package com.leetcode.validSudoku;

public class ValidSudoku {

	public boolean isValidSudoku(char[][] board) {
        if (board == null || board.length == 0) {
            return false;
        }
        
        for (int i = 0; i < board.length; i += 3) {
            for (int j = 0; j < board.length; j += 3) {
                if (!isValidSquare(board, i, j)) {
                    return false;
                }
            }
        }
        
        for (int i = 0; i < board.length; ++i) {
            if (!isValidLine(board, i)) {
                return false;
            }
        }
        
        return true;
    }
    
    private boolean isValidSquare(char[][] board, int rowIndex, int columnIndex) {
        int rowBound = rowIndex + 3;
        int columnBound = columnIndex + 3;
        int[] table = new int[10];
        int digit = 0;
        for (int i = rowIndex; i < rowBound; ++i) {
            for (int j = columnIndex; j < columnBound; ++j) {
                if (board[i][j] != '.') {
                    digit = board[i][j] - '0';
                    if (table[digit] == 0) {
                        table[digit] = 1;
                    } else {
                        return false;
                    }
                }
            }
        }
        
        return true;
    }
    
    private boolean isValidLine(char[][] board, int index) {
        int bound = board.length;
        int[] table = new int[10];
        int digit = 0;
        for (int i = 0; i < bound; ++i) {
            if (board[index][i] != '.') {
                digit = board[index][i] - '0';
                if (table[digit] == 0) {
                    table[digit] = 1;
                } else {
                    return false;
                }
            }
        }
        
        for (int i = 1; i < 10; ++i) {
            table[i] = 0;
        }
        
        for (int i = 0; i < bound; ++i) {
            if (board[i][index] != '.') {
                digit = board[i][index] - '0';
                if (table[digit] == 0) {
                    table[digit] = 1;
                } else {
                    return false;
                }
            }
        }
        
        return true;
    }

}

File: Python/valid_sudoku/valid_sudoku.py
class Solution(object):
	def isValidSudoku(self, board):
		if board == []:
			return False
		for i in range(9):
			s = set()
			for j in range(9):
				if board[i][j] != '.':
					if board[i][j] in s:
						return False
					s.add(board[i][j])

		for j in range(9):
			s = set()
			for i in range(9):
				if board[i][j] != '.':
					if board[i][j] in s:
						return False
					s.add(board[i][j])

		for i in range(3):
			for j in range(3):
				s = set()
				for m in range(3):
					for n in range(3):
						if board[i*3+m][j*3+n] != '.':
							if board[i*3+m][j*3+n] in s:
								return False
							s.add(board[i*3+m][j*3+n])

		return True

